Slim down `Element`, various dtype-related improvements and changes

aldanor opened the pull request:
@adamreichold Here comes, as promised 😄 

This is the preliminary PR to enable further work on #254.

Although there's some breaking changes in here, it's probably better to do it all in one batch because many of these changes are inter-related. I believe all changes are improvements, in one way or another; and there's also some fixes.

An unsorted list of changes:

- Replace `Element::DATA_TYPE` with `Element::IS_POD`
  - Otherwise, we would run into serious complications when trying to implement record types. Also, this is much simpler since all we need really is a quick way to check whether a type is pod or not.
- Remove `Element::same_type()`
  - This was based on typenums and wouldn't work with more complex types. Now it uses `PyArray_EquivTypes`.
- Use `PyArray_NewFromDescr` instead of `PyArray_New` to create arrays.
  - This detaches it from typenums and will allow using custom descriptors later on.
- Added a FIXME note in `FromPyObject::extract()` - it currently doesn't check the instance type and only verifies that dtype is 'O' which may and will lead to unsafe behaviour and so it has to be fixed.
- Split a weird `ShapeError` into `DimensionalityError` and `TypeError`. The latter is no longer typenum-based either and would work with any dtypes; formatting is left for numpy to handle.
- Add methods to `PyArrayDescr`:
  - `into_dtype_ptr()` - an alternative to `as_dtype_ptr()` that increfs it; useful since numpy API often steals descriptor references
  - `of<T>()` - an equivalent to `pybind11::dtype::of<T>()`
  - `is_equiv_to()` - to check if descriptor types are equivalent (`PyArray_EquivTypes`)
  - `get_typenum()` is made public since `DataType::from_typenum()` is public; doesn't make much sense to hide it
  - `object()` - a shortcut for creating 'O' dtype (useful in user implementations of `Element`)
- Add `get_typenum()` to `DataType`
- Fix how integer types are mapped to npy types and added a test for it (which was previously failing):
  - As an example, `DataType::Uint64` could be previously mapped to `np.c_ulonglong` instead of `np.uint64` which is *not* the same thing. Now, `u64` always maps to `np.uint64`.
  - Now it uses the same logic as numpy itself (and same as in pybind11)
  - Reverse conversions (npy -> DataType) have also been reimplemented and cleaned up
- Implemented `Element` for `isize`

Open question: one thing that I find extremely confusing is that `DataType::Complex32` maps to `np.complex64` (and `Complex64` maps to `np.complex128`). Wouldn't it make more sense if they were named consistently? (i.e. same as in numpy, 64/128)

(if this is accepted, I can also work on updating the changelog if needed so.)

aldanor commented:
One thing we've discussed in #254 and I'd like to point out here again:

`IS_POD` is actually somewhat redundant as well because there's [`dtype.hasobject`](https://numpy.org/doc/stable/reference/generated/numpy.dtype.hasobject.html) flag (and related stuff like [`_may_have_objects()`](https://github.com/numpy/numpy/blob/7125cdfa21081b91907a65b6c888791cf150a0b2/numpy/core/src/multiarray/common.c#L373)). The only question is whether we'd like to sacrifice a few nanoseconds to slim down `Element` to just a single `get_dtype()`.

kngwyu commented on src/dtype.rs:
```diff
@@ -217,74 +273,95 @@ impl DataType {
+    /// It should be set to true for all trivially copyable types (like scalar types
+    /// and record/array types only containing trivially copyable fields and elements).
+    ///
+    /// This flag should *always* be set to `true` for object types.
```
!
```suggestion
    /// This flag should *always* be set to `false` for object types.
```

aldanor replied:
Thanks... 🤦

kngwyu commented on src/dtype.rs:
```diff
@@ -217,74 +273,95 @@ impl DataType {
+    /// and record/array types only containing trivially copyable fields and elements).
+    ///
+    /// This flag should *always* be set to `true` for object types.
+    const IS_POD: bool;
```
I like `IS_COPYABLE` or so

aldanor replied:
Here's a few arguments if I might:
- `IS_COPYABLE` is a bit confusing: you can have `#[derive(Copy)] struct Foo {}` type in Rust which you stick into a `Py<Foo>` numpy array that you can `.copy()` in Python.
- It's *not* only about copying. E.g. it may matter when extracting fields out of record arrays, extracting subarrays etc.

Alternative option is `HAS_OBJECT`, I guess...

adamreichold replied:
Yes, I agree. We should stay closer to Rust's `Copy` terminology here. POD might be familiar to C++ users, but it is rarely used in Rust code.

adamreichold replied:
The argument on `struct Foo` applies just as well to POD, as `Foo` is POD, but `Py<Foo>` is not. The point is also not to limit the discussion to copying, but to align the nomenclature to Rust's `Copy` trait (on the element type, which would be `Py<Foo>`, not `Foo`).

aldanor replied:
If we want to keep `IS_POD` / `HAS_OBJECT` around, we could also sprinkle some `debug_assert`s around, at least to check the trivial 'O' dtype case.

adamreichold replied:
You mean `IS_COPY` ;-) But yeah, adding `debug_assert`s checking that this matches the descriptor is certainly a good idea.

aldanor replied:
Ok, I've renamed it to `IS_COPY` and extended the docstring a bit to also mention record types.

kngwyu reviewed (commented):
Thanks, looks nice, though I haven't checked the error types yet

adamreichold commented on src/error.rs:
```diff
@@ -76,7 +24,57 @@ macro_rules! impl_pyerr {
 }
 
-impl_pyerr!(ShapeError);
+/// Represents that dimensionalities of the given arrays don't match.
```
I think "dimensionality" is a unnecessary. Why not just the "dimension of the given arrays" and similarly in the type name?

aldanor replied:
`DimensionError` may be misread as "dimensions mismatch" (I would *definitely* read it as that myself) which would be incorrect because dimensions=shape and we don't check any shapes but rather only `ndim`. Another option is `NdimError` but it's rather ugly (although still an option).

E.g.: [`ndarray::ArrayBase::into_dimensionality()`](https://docs.rs/ndarray/0.15.4/ndarray/struct.ArrayBase.html#method.into_dimensionality)

adamreichold replied:
Fair enough. Let's avoid the confusion.

adamreichold commented on .gitignore:
```diff
@@ -15,3 +15,7 @@ build/
 **/dist/
 __pycache__
+
+# Ignore project files
```
This does not belong into a project-wide gitignore file as it is specific to each developer. You can define a global `core.excludesfile` matching your chosen tool set.

aldanor replied:
I've temporarily added it when starting to work on a branch - thanks, will remove.

adamreichold commented on src/array.rs:
```diff
@@ -136,13 +136,23 @@ impl<'a, T: Element, D: Dimension> FromPyObject<'a> for &'a PyArray<T, D> {
+        let src_dtype = array.dtype();
+        let dst_dtype = T::get_dtype(ob.py());
+        if !src_dtype.is_equiv_to(dst_dtype) {
+            // FIXME: this is unsound - we're only checking that the dtype is 'O'
```
I don't think this is really correct. This is only applies to object types and we are not providing any `Element` impls besides `Py<PyAny>` which is safe. So together with #255, I don't see any soundness issues here?

aldanor replied:
You're implying that it's sound because the `Element` trait is unsafe to impl? It doesn't make it not unsound - we're still technically providing a footgun, no?

adamreichold replied:
> You're implying that it's sound because the Element trait is unsafe to impl?

That is the technical argument, yes.

> It doesn't make it not unsound - we're still technically providing a footgun, no?

It does make it not sound if that is part of the traits contract. In #255, it is expressed that one cannot implement an object-like `Element` for anything but `PyObject` for now as we do not check each elements dynamic type. We might relax this eventually, but as it stands, there is no unsoundness after #255 is applied.

aldanor replied:
Fair enough, I'll remove the note.

It would probably be nice to have other-than-PyAny support at some point, but it's good for now.

adamreichold commented on src/dtype.rs:
```diff
@@ -43,6 +47,13 @@ impl PyArrayDescr {
+    /// Returns `self` as `*mut PyArray_Descr` and transfers the ownership.
+    ///
+    /// Useful in cases where the descriptor is stolen by the API.
+    pub fn into_dtype_ptr(&self) -> *mut PyArray_Descr {
```
But wouldn't it be more helpful if this method were to take `self` by value then?

adamreichold replied:
Ah sorry, from the cover it seems that this does not really "transfer ownership", it just increments the reference count. Could you modify the comment in that direction? Thanks!

aldanor replied:
I guess it's really meant to be `<&'a PyArrayDescr>::into_dtype_ptr()`, but that would be a bit too obscure.

Good point - I'll update the docstring.

adamreichold commented on src/dtype.rs:
```diff
@@ -99,72 +128,99 @@ pub enum DataType {
 }
 
 impl DataType {
```
I wonder if we need to keep `DataType` around at all after this? Couldn't we call the relevant helper functions directly via the (invocations of the) `impl_num_element!` macro and remove this type from our API?

aldanor replied:
This has crossed my mind as well... so what are your concrete suggestions in this case?

(Note that `DataType` has also been removed from error types as a part of migrating `DATA_TYPE` to `IS_POD` which also removes that link)

adamreichold replied:
I'd say drop `enum DataType`, pass the `NPY_*` constants or calls to a module-private function `npy_int_type_lookup` directly to the macro.

aldanor replied:
Looks like it's doable. What we'll lose, I guess, is being able to do something like `dtype.datatype() == Some(DataType::Uint32)`. But the question is, is that really useful? (most likely not)

adamreichold commented:
> Open question: one thing that I find extremely confusing is that DataType::Complex32 maps to np.complex64 (and Complex64 maps to np.complex128). Wouldn't it make more sense if they were named consistently? (i.e. same as in numpy, 64/128)

I guess this would be best resolved by dropping `DataType` entirely?

aldanor commented:
> 



> > Open question: one thing that I find extremely confusing is that DataType::Complex32 maps to np.complex64 (and Complex64 maps to np.complex128). Wouldn't it make more sense if they were named consistently? (i.e. same as in numpy, 64/128)
> 
> I guess this would be best resolved by dropping `DataType` entirely?

That's a separate question, although a good point. I'll take a look if we can drop it now.

What about type aliases?

```rust
c32 <--> complex64
c64 <--> complex128
```

adamreichold commented:
> What about type aliases?

I think we should leave them, they are consistent with the nomenclature of the `num-complex` crate which is what Rust code would use to actually work with complex numbers. Both perspectives, i.e. size of the aggregate and size of the components, seem understandable to me, and hence this feels like unnecessary breakage that just moves the confusion elsewhere but does not remove it.

aldanor commented:
> > What about type aliases?
> 
> I think we should leave them, they are consistent with the nomenclature of the `num-complex` crate which is what Rust code would use to actually work with complex numbers. Both perspectives, i.e. size of the aggregate and size of the components, seem understandable to me, and hence this feels like unnecessary breakage that just moves the confusion elsewhere but does not remove it.

Maybe, at the very least then, it's worth mentioning it in the docstrings of type aliases? (that c32 is `np.complex64` etc).

adamreichold commented:
> Maybe, at the very least then, it's worth mentioning it in the docstrings of type aliases? (that c32 is np.complex64 etc).

Actually, I have considered this further, we should drop the outright and just use `Complex32/64`. Is not significantly more typing and it makes it clear that this refers to the types from `num-complex` and not some bespoke binding type.

The doc comment would make sense though, but I would then attach it to the trait impl for these two types (as the type aliases are no more).

kngwyu commented:
I'm also OK to remove `c32/c64` aliases

kngwyu commented on src/error.rs:
```diff
@@ -76,7 +24,57 @@ macro_rules! impl_pyerr {
+/// Represents that types of the given arrays don't match.
+#[derive(Debug)]
+pub struct TypeError {
+    from: String,
+    to: String,
```
I personally like having `Option<String>` and unwrapping it when formatting, but it's too nitpicking and either is fine

aldanor replied:
I think in cases like this, having `Option<String>` may indicate that there may be type information available (i.e., logically available) or not, whereas here we always have both from and to types.

So I'd rather leave it like this if it's ok.

kngwyu reviewed (changes requested):
Thank you @aldanor and @adamreichold, I like the way this PR simplified things

aldanor commented:
> I'm also OK to remove `c32/c64` aliases

Already done (`Complex32` and `Complex64` re-exports remain though; with extra docstring in their `Element` impl mentioning that they map to `np.complex64` and `np.complex128` respectively, to avoid confusion).

aldanor commented:
@adamreichold @kngwyu I'm almost done with dropping `DataType` entirely as suggested by @adamreichold - this simplifies things even further which looks good.

One (cosmetic) question though – do we want a top-level `dtype::<T>` or `dtype_of::<T>()` function perhaps? That would be a lot more obvious to newcomers than having to dig into documentation for `PyArrayDescr` (I bet most sane people who use numpy haven't ever heard the 'array descr' term, but everyone knows what dtype is, so having a function at crate level might be nice and friendly).

This came up from fixing multiple doctests that used to look like this:

```rust
assert_eq!(dtype.get_datatype().unwrap(), numpy::DataType::Int32);
```

The current version is

```rust
assert_eq!(dtype.is_equiv_to(numpy::PyArrayDescr::of::<i32>(py)));
```

My suggestion is to also add a top-level alias, like
```rust
assert_eq!(dtype.is_equiv_to(numpy::dtype::<i32>(py)));
// or
assert_eq!(dtype.is_equiv_to(numpy::dtype_of::<i32>(py)));
```

aldanor commented on src/dtype.rs:
```diff
@@ -217,74 +272,99 @@ impl DataType {
-    } else if #[cfg(all(target_pointer_width = "32", windows))] {
-            impl_num_element!(usize, Uint32, NPY_UINT, NPY_ULONG);
-    } else if #[cfg(all(target_pointer_width = "32", not(windows)))] {
-            impl_num_element!(usize, Uint32, NPY_UINT);
+    if #[cfg(any(target_pointer_width = "32", target_pointer_width = "64"))] {
+        impl_num_element!(usize, DataType::integer::<usize>().unwrap());
+        impl_num_element!(isize, DataType::integer::<isize>().unwrap());
```
That crossed my mind as well. I'll take a look at dropping it

aldanor replied:
Ok, `cfg_if` dependency has been removed now.

adamreichold commented:
(You can check the extensions by running `cargo check/clippy --workspace` locally.)

aldanor commented:
> (You can check the extensions by running `cargo check/clippy --workspace` locally.)

Yea, totally forgot about the `--workspace`. That's fixed now. (re: clippy - I've left a comment on #239)

adamreichold commented on src/dtype.rs:
```diff
@@ -217,74 +155,142 @@ impl DataType {
+    ($ty:ty => $npy_type:ident $(,#[$meta:meta])*) => {
+        impl_element_scalar!(@impl: $ty, NPY_TYPES::$npy_type $(,#[$meta])*);
+    };
+    ($ty:ty) => {
```
Couldn't these two arm be merged via using `$($tys:ty),*` below?

aldanor replied:
Good catch, missed that.

aldanor commented on src/dtype.rs:
```diff
@@ -217,74 +272,99 @@ impl DataType {
+                    assert_eq!(type_name::<usize>(py), "uint64");
+                    assert_eq!(type_name::<isize>(py), "int64");
+                }
+                _ => {}
```
Fair enough (matter of preference though) - fixed

aldanor commented:
I think all of the questions, suggestions and concerns raised so far have been addressed and it should be good to go.

Should the unreleased section of the changelog be updated as well?

adamreichold commented:
> Should the unreleased section of the changelog be updated as well?

Technically, this is breaking and hence should be mentioned. However, the only way a user of this crate could use `Element` until know was for trivial wrappers of primitive types, so I don't think this is actually used anywhere.

aldanor commented:
Rebased on the latest master and updated the changelog (there's actually quite a few various public-facing changes, not just the `Element` change itself).

adamreichold commented:
Is anyone still planning to review this but hasn't had the time so far? If not, I would like merge this on the weekend as I think it is a definite improvement by simplification and has two approvals already.
